File: 21.合并两个有序链表.cpp
// https://leetcode-cn.com/problems/merge-two-sorted-lists/
#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode* mergeTwoLists(ListNode* l1, ListNode* l2);

int main()
{
	return 0;
}

ListNode* mergeTwoLists(ListNode* l1, ListNode* l2)
{
	ListNode *head, *end;
	
	head = new ListNode(-1);
	end = head;
	while(l1!=nullptr && l2!=nullptr)
	{
		if(l1->val < l2->val)
		{
			end->next = l1;
			l1 = l1->next;
		}
		else
		{
			end->next = l2;
			l2 = l2->next;
		}
		end = end->next;
	}
	
	end->next = (l1 == nullptr) ? l2 : l1;
	
	return head->next;
}

File: 链表/21.合并两个有序链表.java
// https://leetcode.cn/problems/merge-two-sorted-lists/
// 题解：新建一个链表头进行排序。
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode ans = new ListNode(-1);
        ListNode temp = ans;

        while(list1 != null && list2 != null) {
            if(list1.val <= list2.val) {
                temp.next = list1; list1 = list1.next;
            } else {
                temp.next = list2; list2 = list2.next;
            }
            temp = temp.next;
        }
        temp.next = (list1 == null) ? list2 : list1;

        return ans.next;
    }
}

/*
[1,2,4]
[1,3,4]

[1,1,2,3,4,4]
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */